
sys/conf and sys/sys/cdefs.h bits submitted (as separate patches) to
freebsd-hackers on 2011-07-12 and 2011-07-04, respectively.

Rest of the patch needs some cleanup before it can be submitted.

--- a/sys/conf/kern.pre.mk
+++ b/sys/conf/kern.pre.mk
@@ -90,6 +90,14 @@
 .endif
 WERROR?= -Werror
 
+.if !defined(OPSYS)
+OPSYS!= uname -s
+.endif
+
+.if ${OPSYS} != "FreeBSD"
+CFLAGS+= -Ulinux -U__linux__ -D__FreeBSD__
+.endif
+
 # XXX LOCORE means "don't declare C stuff" not "for locore.s".
 ASM_CFLAGS= -x assembler-with-cpp -DLOCORE ${CFLAGS}
 
--- a/sys/conf/kmod.mk
+++ b/sys/conf/kmod.mk
@@ -96,6 +96,14 @@
 CFLAGS+=	-D_KERNEL
 CFLAGS+=	-DKLD_MODULE
 
+.if !defined(OPSYS)
+OPSYS!=		uname -s
+.endif
+
+.if ${OPSYS} != "FreeBSD"
+CFLAGS+=	-Ulinux -U__linux__ -D__FreeBSD__
+.endif
+
 # Don't use any standard or source-relative include directories.
 CSTD=		c99
 NOSTDINC=	-nostdinc
--- a/sys/contrib/ipfilter/netinet/ip_compat.h
+++ b/sys/contrib/ipfilter/netinet/ip_compat.h
@@ -985,10 +985,10 @@
 typedef struct mbuf mb_t;
 # endif /* _KERNEL */
 
-# if __FreeBSD__ < 3
+# if 0
 #  include <machine/spl.h>
 # else
-#  if __FreeBSD__ == 3
+#  if 0
 #   if defined(IPFILTER_LKM) && !defined(ACTUALLY_LKM_NOT_KERNEL)
 #    define	ACTUALLY_LKM_NOT_KERNEL
 #   endif
--- a/sys/contrib/ipfilter/netinet/ip_proxy.c
+++ b/sys/contrib/ipfilter/netinet/ip_proxy.c
@@ -63,7 +63,7 @@
 # include <sys/stream.h>
 # include <sys/kmem.h>
 #endif
-#if __FreeBSD__ > 2
+#if 1
 # include <sys/queue.h>
 #endif
 #include <net/if.h>
--- a/sys/dev/bktr/bktr_reg.h
+++ b/sys/dev/bktr/bktr_reg.h
@@ -717,7 +717,7 @@
 /* ioctl_cmd_t int on old versions, u_long on new versions */
 /***********************************************************/
 
-#if (__FreeBSD__ == 2)
+#if 0
 typedef int ioctl_cmd_t;
 #endif
 
--- a/sys/net/if_spppfr.c
+++ b/sys/net/if_spppfr.c
@@ -25,7 +25,7 @@
 
 #include <sys/param.h>
 
-#if defined(__FreeBSD__) && __FreeBSD__ >= 3
+#if defined(__FreeBSD__)
 #include "opt_inet.h"
 #include "opt_inet6.h"
 #include "opt_ipx.h"
@@ -45,7 +45,7 @@
 #include <sys/sockio.h>
 #include <sys/socket.h>
 #include <sys/syslog.h>
-#if defined(__FreeBSD__) && __FreeBSD__ >= 3
+#if defined(__FreeBSD__)
 #include <sys/random.h>
 #endif
 #include <sys/malloc.h>
@@ -149,7 +149,7 @@
 	unsigned short  ptarget2;
 } __packed;
 
-#if defined(__FreeBSD__) && __FreeBSD__ >= 3 && __FreeBSD_version < 501113
+#if defined(__FreeBSD__) && __FreeBSD_version < 501113
 #define	SPP_FMT		"%s%d: "
 #define	SPP_ARGS(ifp)	(ifp)->if_name, (ifp)->if_unit
 #else
--- a/sys/sys/cdefs.h
+++ b/sys/sys/cdefs.h
@@ -349,7 +349,7 @@
 #endif
 
 /* Compiler-dependent macros that rely on FreeBSD-specific extensions. */
-#if __FreeBSD_cc_version >= 300001 && defined(__GNUC__) && !defined(__INTEL_COMPILER)
+#if defined(__FreeBSD_cc_version) && __FreeBSD_cc_version >= 300001 && defined(__GNUC__) && !defined(__INTEL_COMPILER)
 #define	__printf0like(fmtarg, firstvararg) \
 	    __attribute__((__format__ (__printf0__, fmtarg, firstvararg)))
 #else
